MatrixSDK/VoIP/CallKit/MXCallKitAdapter.m

```diff
@@ -0,0 +1,285 @@
+    update.remoteHandle = handle;
+    update.localizedCallerName = caller.displayname;
+    update.hasVideo = call.isVideoCall;
+    update.supportsHolding = NO;
+    update.supportsGrouping = NO;
+    update.supportsUngrouping = NO;
+    update.supportsDTMF = NO;
+    
+    [self.provider reportNewIncomingCallWithUUID:callUUID update:update completion:^(NSError * _Nullable error) {
+        if (error)
+        {
+            [call hangup];
+            return;
+        }
+        
+        self.calls[callUUID] = call;
+        
+#ifdef MX_CALL_STACK_JINGLE
+        // Workaround from https://forums.developer.apple.com/message/169511 suggests configuring audio in the
+        // completion block of the `reportNewIncomingCallWithUUID:update:completion:` method instead of in
+        // `provider:performAnswerCallAction:` per the WWDC examples.
+        [MXJingleCallAudioSessionConfigurator configureAudioSessionForVideoCall:call.isVideoCall];
+#endif
+    }];
+    
+}
+
+- (void)reportCall:(MXCall *)call startedConnectingAtDate:(nullable NSDate *)date
+{
+    [self.provider reportOutgoingCallWithUUID:call.callUUID startedConnectingAtDate:date];
+}
+
+- (void)reportCall:(MXCall *)call connectedAtDate:(nullable NSDate *)date
+{
+    [self.provider reportOutgoingCallWithUUID:call.callUUID connectedAtDate:date];
+}
+
++ (BOOL)callKitAvailable
+{
+    return [NSProcessInfo.processInfo isOperatingSystemAtLeastVersion:(NSOperatingSystemVersion){10,0,0}];
+}
+
+#pragma mark - CXProviderDelegate
+
+- (void)providerDidReset:(CXProvider *)provider
+{
+    NSLog(@"Provider did reset");
+}
+
+- (void)provider:(CXProvider *)provider didActivateAudioSession:(AVAudioSession *)audioSession
+{
+#ifdef MX_CALL_STACK_JINGLE
+    [MXJingleCallAudioSessionConfigurator audioSessionDidActivate:audioSession];
+#endif
+}
+
+- (void)provider:(CXProvider *)provider didDeactivateAudioSession:(AVAudioSession *)audioSession
+{
+#ifdef MX_CALL_STACK_JINGLE
+    [MXJingleCallAudioSessionConfigurator audioSessionDidDeactivate:audioSession];
+#endif
+}
+
+- (void)provider:(CXProvider *)provider performStartCallAction:(CXStartCallAction *)action
+{
+    MXCall *call = self.calls[action.callUUID];
+    if (!call)
+    {
+        [action fail];
+        return;
+    }
+    
+#ifdef MX_CALL_STACK_JINGLE
+    [MXJingleCallAudioSessionConfigurator configureAudioSessionForVideoCall:call.isVideoCall];
+#endif
+    
+    [action fulfill];
+}
+
+- (void)provider:(CXProvider *)provider performAnswerCallAction:(CXAnswerCallAction *)action
+{
+    MXCall *call = self.calls[action.callUUID];
+    if (!call)
+    {
+        [action fail];
+        return;
+    }
+    
+    [call answer];
+    
+    [action fulfill];
+}
+
+- (void)provider:(CXProvider *)provider performEndCallAction:(CXEndCallAction *)action
+{
+    MXCall *call = self.calls[action.callUUID];
+    if (!call)
+    {
+        [action fail];
+        return;
+    }
+    
+    [call hangup];
+    self.calls[action.UUID] = nil;
```

Comment:
It doesn't matter for me what syntax to use, but using subscripting operator looks less verbos e than calling the method. I really want you to dispel your fear about this by having a look at Clang (documentation)[https://clang.llvm.org/docs/ObjectiveCLiterals.html#dictionary-style-subscripting]. For `NSMutableDictionary` using subscripting operator for setting values is equal to calling `setObject:forKeyedSubscript:` and it is specified in the documentation that **passing nil will cause any object corresponding to aKey to be removed from the dictionary**. So there is no reason to fear about it;)

But an exception is the case when we are using the new literal syntax for creating dictionaries. If the key or the value is `nil` then we will have a runtime error.

I just want to clarify the situation. It's for you to decide which way to use.